addon/bootstrap.js

```diff
@@ -1,144 +1,77 @@
-    studyUtils.firstSeen(); // sends telemetry "enter"
-    const eligible = await config.isEligible(); // addon-specific
-    if (!eligible) {
-      // uses config.endings.ineligible.url if any,
-      // sends UT for "ineligible"
-      // then uninstalls addon
-      await studyUtils.endStudy({reason: "ineligible"});
+this.Bootstrap = {
+  install() {
+  },
+
+  /**
+   * @param addonData Array [ "id", "version", "installPath", "resourceURI", "instanceID", "webExtension" ]
+   * @param reason
+   * @returns {Promise.<void>}
+   */
+  async startup(addonData, reason) {
+    // Check if the user is opted in to pioneer and if not end the study
+    await Pioneer.startup(addonData);
+    const events = Pioneer.utils.getAvailableEvents();
+
+    const isEligible = await Pioneer.utils.isUserOptedIn();
+    if (!isEligible) {
+      console.log('Not eligable for Pioneer study. Will uninstall the study add-on.');
+      Pioneer.utils.endStudy(events.INELIGIBLE);
       return;
     }
-  }
-
-  // for all 'eligible' users, startup.
-  await studyUtils.startup({reason});
-
-  // log what the study variation and other info is.
-  console.log(`info ${JSON.stringify(studyUtils.info())}`);
-
-
-  // if you have code to handle expiration / long-timers, it could go here
-
 
-  // If your study has an embedded webExtension, start it.
-  const webExtension = addonData.webExtension;
-  if (webExtension) {
-    webExtension.startup().then(api => {
-      const {browser} = api;
-      /* spec for messages intended for Shield =>
-        {shield:true,msg=[info|endStudy|telemetry],data=data}
-      */
-      browser.runtime.onMessage.addListener(studyUtils.respondToWebExtensionMessage);
-
-      // other browser.runtime.onMessage handlers for your addon, if any
-
-    });
-  }
-
-  // Fire this once (only during INSTALL), then we are done.
-  if (REASONS[reason] === "ADDON_INSTALL" || REASONS[reason] === "ADDON_UPGRADE") {
-    new StudyTelemetryCollector(studyUtils, variation).start();
-  }
-}
-
-
-function shutdown(addonData, reason) {
-  console.log("shutdown", REASONS[reason] || reason);
-  // FRAGILE: handle uninstalls initiated by USER or by addon
-  if (reason === REASONS.ADDON_UNINSTALL || reason === REASONS.ADDON_DISABLE) {
-    console.log("uninstall or disable");
-    if (!studyUtils._isEnding) {
-      // we are the first 'uninstall' requestor => must be user action.
-      console.log("user requested shutdown");
-      studyUtils.endStudy({reason: "user-disable"});
-      return;
+    // Fire this once per released version (only during INSTALL or UPGRADE), then we are done.
+    if (reason === REASONS.ADDON_INSTALL || reason === REASONS.ADDON_UPGRADE) {
+      new StudyTelemetryCollector(Pioneer.utils).start();
     }
-    // normal shutdown, or 2nd uninstall request
-    console.log("Jsms unloading");
-
+  },
 
-    // QA NOTE:  unload addon specific modules here.
+  shutdown() {
+    Cu.unload("resource://esper-pioneer-shield-study/Config.jsm");
```

Comment:
You need to unload all the resources loaded by any of these modules, like `Helpers.jsm`.

Reply:
Thanks, I thought it was only necessary to unload the ones loaded in bootstrap.js - fixed now :)